plugin-api/src/main/java/org/sonar/api/batch/sensor/issue/NewIssueLocation.java

```diff
@@ -53,4 +54,9 @@ public interface NewIssueLocation {
    */
   NewIssueLocation message(String message);
 
+  /**
+   * Optional, parts of message can be treated specially in the frontend
+   * */
+  NewIssueLocation messageHighlighting(List<MessageHighlighting> messageHighlighting);
```

Comment:
wouldn't it be good to have a factory method for `MessageHighlighting`?

Reply:
or just `newIssueLocation.formatMessage(1, 2, Type.CODE)`, unless there's a need to pass a list of pre-computed offsets.

Reply:
Are you suggesting to only provide the `.formatMessage(1, 2, Type.CODE)` interface and not the one accepting a `List<MessageHighlighting>`? Ultimately we want to avoid devs needing to manually specify ranges, as that is very error-prone. So in either case, we would need to write a wrapper, either using some templating or internal formatting scheme, and/or using a builder pattern. Since we can only pass a single message to `NewIssueLocation` anyway, we'll have to do some aggregation before calling `.message(...)` while transforming whatever format the syntactic-sugar layer provides to a plaintext message + formatting anyway. So then we may as well do aggregation of formatting as well.

It goes back to the fact that the raw API is a bit clunky. To really make it usable, it requires more than just a simple `NewIssueLocation#formatMessage` helper. So while we can build helpers around both suggested APIs, neither will avoid the need for them and I guess we don't want to implement the full helpers in the API. So the question is, do we really want to provide a helper that makes it look a little less clunky but ultimately, we will still need to wrap it, or simply provide the raw APIs, which allow the helpers to interface with them as directly as possible?

Reply:
Also something to consider, the API is not made only for internal usage, it is also for third-party plugin developers. If we consider the API we suggest is error-prone and a helper is very likely needed, I feel a bit concerned about the current proposal. Should the helper you mention be made available to those developers somehow?

Reply:
We decided during the meeting to not implement the helper method for now. It can be added later if needed.